elcid/views.py

```diff
@@ -100,3 +104,105 @@ def get_context_data(self, *args, **kwargs):
             pass
 
         return ctd
+
+
+def ward_sort_key(ward):
+    """"
+    Example wards are
+            "8 West",
+            "PITU",
+            "ICU 4 East",
+            "9 East",
+            "8 South",
+            "9 North",
+            "ICU 4 West",
+            "12 West",
+
+    We want to order them first by number
+    then with the wards that are strings.
+    e.g. ICU
+    """
+
+    start = ward[:2].strip()
+    rest = ward[2:]
+    return (len(start), start, rest)
+
+
+class RenalHandover(LoginRequiredMixin, TemplateView):
+    template_name = "elcid/renal_handover.html"
+
+    def get_unit_ward(self, location):
+        if location.ward and location.bed:
+            return "{}/{}".format(
+                location.ward, location.bed
+            )
+
+    def get_context_data(self, *args, **kwargs):
+        """
+        An ordered dictionary of ward to patient
+
+        patient here is...
+        their name
+        their hospital number
+        their unit/ward
+        that episodes diagnosis
+        all of the clinical advice related to the patients infection service
+        """
+        ctx = super().get_context_data(*args, **kwargs)
+        episodes = Renal().get_queryset().order_by("-episode_id")
+        episodes = episodes.order_by("-start")
+        episodes = episodes.prefetch_related(
+            "location_set", "diagnosis_set"
+        )
+        by_ward = defaultdict(list)
+
+        # We don't want to show duplicate episodes in the case of
+        # multiple episodes of the same patient added to the renal list.
+        #
+        # So we just show the most recent episode per patient.
+        patient_ids = set()
+        for episode in episodes:
+            if episode.patient_id in patient_ids:
+                continue
+            patient_ids.add(episode.patient_id)
+            location = episode.location_set.all()[0]
+            diagnosis = ", ".join(
+                i.condition for i in episode.diagnosis_set.all()
+            )
+            demographics = models.Demographics.objects.filter(
+                patient__episode=episode
+            ).get()
+            microbiology_inputs = models.MicrobiologyInput.objects.filter(
+                episode__patient__episode=episode
```

Comment:
Is this definitely right ?
Does it not then put us back in the same episodic advice situation?

Reply:
So we do give me inner join on patient then inner join on the episode table again which is basically expected, but as discussed, let just get the patient and that can save lots of joins.

```
SELECT "elcid_microbiologyinput"."id", "elcid_microbiologyinput"."created", "elcid_microbiologyinput"."updated", "elcid_microbiologyinput"."created_by_id", "elcid_microbiologyinput"."updated_by_id", "elcid_microbiologyinput"."consistency_token", "elcid_microbiologyinput"."episode_id", "elcid_microbiologyinput"."when", "elcid_microbiologyinput"."initials", "elcid_microbiologyinput"."infection_control", "elcid_microbiologyinput"."clinical_discussion", "elcid_microbiologyinput"."agreed_plan", "elcid_microbiologyinput"."discussed_with", "elcid_microbiologyinput"."clinical_advice_given", "elcid_microbiologyinput"."infection_control_advice_given", "elcid_microbiologyinput"."change_in_antibiotic_prescription", "elcid_microbiologyinput"."referred_to_opat", "elcid_microbiologyinput"."white_cell_count", "elcid_microbiologyinput"."c_reactive_protein", "elcid_microbiologyinput"."maximum_temperature", "elcid_microbiologyinput"."liver_function_fk_id", "elcid_microbiologyinput"."liver_function_ft", "elcid_microbiologyinput"."reason_for_interaction_fk_id", "elcid_microbiologyinput"."reason_for_interaction_ft", "elcid_microbiologyinput"."renal_function_fk_id", "elcid_microbiologyinput"."renal_function_ft" FROM 
"elcid_microbiologyinput" INNER JOIN "opal_episode" ON ("elcid_microbiologyinput"."episode_id" = "opal_episode"."id") INNER JOIN "opal_patient" ON ("opal_episode"."patient_id" = "opal_patient"."id") INNER JOIN "opal_episode" T4 ON ("opal_patient"."id" = T4."patient_id") WHERE T4."id" = 12 ORDER BY "elcid_microbiologyinput"."when" DESC; args=(12,)
```

Reply:
we actually have episode.patient_id so we can remove the join with no actual cost. updated.